src/Microsoft.AspNet.OData.Shared/Builder/IODataInstanceAnnotationContainer.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    /// <summary>
+    /// Interface to used as a Container for holding Instance Annotations, An default implementation is provided
+    /// Custoer can implement the interface and can have their own implementation.
+    /// </summary>
+    public interface IODataInstanceAnnotationContainer
+    {
+        /// <summary>
+        /// Method to Add an Instance Annotation to the CLR type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        void AddResourceAnnotation(string annotationName, object value);
```

Comment:
>void AddResourceAnnotation(string annotationName, object value); [](start = 8, length = 64)

All these methods can be extension methods on the interface, Then we can make the interface more simpler. what's your thought @mikepizzo.

Reply:
What would be the interface that we're adding extensions to? Just IDictionary<string,IDictionary<string>>?   I like having the interface as a strongly typed marker interface. It's pretty trivial to implement, and I'm not sure any simpler interface would necessarily be as efficient or as clear (i.e., use of null or empty string to distinguish between the resource and a property on the resource). 

---
In reply to: [490448156](https://github.com/OData/WebApi/pull/2219#discussion_r490448156) [](ancestors = 490448156)

Reply:
Like below: 
    
public static class IODataInstanceAnnotationContainerExtensions
    {
        public static void AddResourceAnnotation(this IODataInstanceAnnotationContainer container, string annotationName, object value)
        {
            if (container == null)
            {
                throw Error.ArgumentNull(nameof(container));
            }
            ...
        }

---
In reply to: [490466288](https://github.com/OData/WebApi/pull/2219#discussion_r490466288) [](ancestors = 490466288,490448156)